timemachine/cpp/src/local_md_potentials.cu

```diff
@@ -0,0 +1,190 @@
+#include "constants.hpp"
+#include "gpu_utils.cuh"
+#include "kernel_utils.cuh"
+#include "kernels/k_flat_bottom_bond.cuh"
+#include "kernels/k_indices.cuh"
+#include "kernels/k_local_md.cuh"
+#include "local_md_potentials.hpp"
+#include "math_utils.cuh"
+#include <cub/cub.cuh>
+#include <random>
+#include <vector>
+
+namespace timemachine {
+
+// Struct to as a CUB <= operation
+struct LessThan {
+    int compare;
+    CUB_RUNTIME_FUNCTION __device__ __forceinline__ explicit LessThan(int compare) : compare(compare) {}
+    CUB_RUNTIME_FUNCTION __device__ __forceinline__ bool operator()(const int &a) const { return (a < compare); }
+};
+
+LocalMDPotentials::LocalMDPotentials(const int N, const std::vector<std::shared_ptr<BoundPotential>> bps)
+    : N_(N), temp_storage_bytes_(0), all_potentials_(bps), restraints_(N_ * 2), bond_params_(N_ * 3),
+      probability_buffer_(round_up_even(N_)), d_free_idxs_(N_), d_row_idxs_(N_), d_col_idxs_(N_), p_num_selected_(1),
+      num_selected_buffer_(1) {
+
+    std::vector<std::shared_ptr<BoundPotential>> nonbonded_pots;
+    get_nonbonded_all_pair_potentials(bps, nonbonded_pots);
+
+    if (nonbonded_pots.size() > 1) {
+        throw std::runtime_error("found multiple NonbondedAllPairs potentials");
+    }
+    if (nonbonded_pots.size() != 1) {
+        throw std::runtime_error("unable to find a NonbondedAllPairs potential");
+    }
+
+    // Only used to reference shared_ptr to potential and for Nonbonded parameters
+    // modifications to the BoundPotential has no impact
+    nonbonded_bp_ = nonbonded_pots[0];
+
+    // Ensure that we allocate enough space for all potential bonds
+    std::vector<int> default_bonds(2 * N_);
+    for (int i = 0; i < N_; i++) {
+        default_bonds[i * 2 + 0] = 0;
+        default_bonds[i * 2 + 1] = i + 1;
+    }
+    restraint_ = std::shared_ptr<FlatBottomBond<double>>(new FlatBottomBond<double>(default_bonds));
+    // Construct a bound potential with 0 params
```

Comment:
did we not want to allocate some default params (similar to the above for bond idxs)?

Reply:
nm - we're deferring this until we remove the ability to modify shapes in bound potentials?